itlwm/hal_iwn/ItlIwn.cpp

```diff
@@ -0,0 +1,7306 @@
+
+#define DEVNAME(_s)    ((_s)->sc_dev.dv_xname)
+
+#define IC2IFP(_ic_) (&(_ic_)->ic_if)
+
+#define IWN_DEBUG
+
+#ifdef IWN_DEBUG
+#define DPRINTF(x)    do { if (iwn_debug > 0) XYLog x; } while (0)
+#define DPRINTFN(n, x)    do { if (iwn_debug >= (n)) XYLog x; } while (0)
+int iwn_debug = 1;
+#else
+#define DPRINTF(x)    do { ; } while (0)
+#define DPRINTFN(n, x)    do { ; } while (0)
+#endif
+
+#define abs(N) ((N<0)?(-N):(N))
+#define M_DEVBUF 2
+
+#ifdef DELAY
+#undef DELAY
+#define DELAY IODelay
+#endif
+
+bool ItlIwn::attach(IOPCIDevice *device)
+{
+    pci.pa_tag = device;
+    pci.workloop = getMainWorkLoop();
+    if (!iwn_attach(&com, &pci)) {
+        detach(device);
+        releaseAll();
+        return false;
+    }
+    return true;
+}
+
+void ItlIwn::
+detach(IOPCIDevice *device)
+{
+    struct _ifnet *ifp = &com.sc_ic.ic_ac.ac_if;
+    ieee80211_ifdetach(ifp);
+    taskq_destroy(systq);
+    releaseAll();
+}
+
+void ItlIwn::
+releaseAll()
+{
+    pci_intr_handle *intrHandler = com.ih;
+    
+    if (intrHandler) {
+        if (intrHandler->intr && intrHandler->workloop) {
+//            intrHandler->intr->disable();
+            intrHandler->workloop->removeEventSource(intrHandler->intr);
+            intrHandler->intr->release();
+        }
+        intrHandler->intr = NULL;
+        intrHandler->workloop = NULL;
+        intrHandler->arg = NULL;
+        intrHandler->dev = NULL;
+        intrHandler->func = NULL;
+        intrHandler->release();
+        com.ih = NULL;
+    }
+    pci.pa_tag = NULL;
+    pci.workloop = NULL;
+}
+
+void ItlIwn::free()
+{
+    XYLog("%s\n", __FUNCTION__);
+    super::free();
+}
+
+ItlDriverInfo *ItlIwn::
+getDriverInfo()
+{
+    return this;
+}
+
+ItlDriverController *ItlIwn::
+getDriverController()
+{
+    return this;
+}
+
+IOReturn ItlIwn::enable(IONetworkInterface *netif)
+{
+    XYLog("%s\n", __FUNCTION__);
+    struct _ifnet *ifp = &com.sc_ic.ic_ac.ac_if;
+    ifp->if_flags |= IFF_UP;
+    iwn_activate(&com, DVACT_WAKEUP);
+    return kIOReturnSuccess;
+}
+
+IOReturn ItlIwn::disable(IONetworkInterface *netif)
+{
+    XYLog("%s\n", __FUNCTION__);
+    iwn_activate(&com, DVACT_QUIESCE);
+    return kIOReturnSuccess;
+}
+
+void ItlIwn::
+clearScanningFlags()
+{
+    com.sc_flags &= ~(IWN_FLAG_SCANNING | IWN_FLAG_BGSCAN);
+}
+
+const char *ItlIwn::
+getFirmwareVersion()
+{
+    return com.fwname;
+}
+
+const char *ItlIwn::
+getFirmwareName()
+{
+    return com.fwname;
+}
+
+UInt32 ItlIwn::
+supportedFeatures()
+{
+    return 0;
+}
+
+const char *ItlIwn::
+getFirmwareCountryCode()
+{
+    return "ZZ";
+}
+
+uint32_t ItlIwn::
+getTxQueueSize()
+{
+    return IWN_TX_RING_COUNT;
+}
+
+int16_t ItlIwn::
+getBSSNoise()
+{
+    return com.noise;
+}
+
+bool ItlIwn::
+is5GBandSupport()
+{
+    return com.sc_flags & IWN_FLAG_HAS_5GHZ;
+}
+
+int ItlIwn::
+getTxNSS()
+{
+    // return !com.sc_nvm.sku_cap_mimo_disable ? (iwn_mimo_enabled(&com) ? 2 : 1) : 1;
+    return 1;
```

Comment:
should not hard coded to 1, some devices may support MIMO, should return the value of `sc->ntxchains`

Reply:
Thanks a lot to point out. I understand incorrectly the MIMO was introduced only to 802.11ac, and it was actually in the N standard. I can find the logs indicate 2T2R, so the TxNSS should be 2.
I also noticed the upload speed increased to 30mbps while test using iperf3 after this change.